src/engine/subscription/Subscription.spec.ts

```diff
@@ -0,0 +1,575 @@
+              required: true,
+            },
+          },
+          subscriptions: [
+            new Subscription({
+              type: SUBSCRIPTION_TYPE_CREATE,
+              name: 'onBuild',
+              description: 'build item',
+              attributes: ['someAttribute'],
+            }),
+          ],
+        });
+
+        otherEntity.getSubscriptionByName('onBuild');
+      }
+
+      expect(fn).toThrowErrorMatchingSnapshot();
+    });
+
+    it('should throw on duplicate subscription names', () => {
+      const subscriptions = [
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_CREATE,
+          name: 'onBuild',
+          description: 'build item',
+          attributes: ['someAttribute'],
+        }),
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_DELETE,
+          name: 'onBuild',
+          description: 'build item',
+          attributes: ['someAttribute'],
+        }),
+      ];
+
+      function fn() {
+        processEntitySubscriptions(entity, subscriptions);
+      }
+
+      expect(fn).toThrowErrorMatchingSnapshot();
+    });
+
+    it('should throw if unknown attributes are used', () => {
+      const subscriptions = [
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_CREATE,
+          name: 'onBuild',
+          description: 'build item',
+          attributes: ['doesNotExist'],
+        }),
+      ];
+
+      function fn() {
+        processEntitySubscriptions(entity, subscriptions);
+      }
+
+      expect(fn).toThrowErrorMatchingSnapshot();
+    });
+
+    it('should allow for empty attribute lists on DELETE type subscriptions', () => {
+      const subscriptions = [
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_DELETE,
+          name: 'onDrop',
+          description: 'drop item',
+          attributes: [],
+        }),
+      ];
+
+      processEntitySubscriptions(entity, subscriptions);
+    });
+  });
+
+  describe('preProcessor', () => {
+    const testEntity = new Entity({
+      name: 'SomeTestsEntityName',
+      description: 'Just some description',
+      attributes: {
+        someAttribute: {
+          type: DataTypeString,
+          description: 'Just some description',
+          required: false,
+        },
+        anotherAttribute: {
+          type: DataTypeString,
+          description: 'Just some description',
+        },
+      },
+      subscriptions: [
+        new Subscription({
+          name: 'SomeSubWithPreProcessor',
+          type: SUBSCRIPTION_TYPE_CREATE,
+          description: 'build item',
+          attributes: ['someAttribute'],
+          delimiter: '/',
+          // wildCard: '',
+          // pattern: '',
+          preProcessor: (
+            _entity,
+            _id,
+            _source,
+            input,
+            typeName,
+            entitySubscription,
+          ) => {
+            if (entitySubscription.attributes && Object.keys(input).length) {
```

Comment:
Here we are allowed to use the preProcessor to generate a custom topic that will be used by PubSub instance in `getSubscriptionResolver` .
The default topic being `${entitySubscription.name}${entity.name}`